doc/sphinx/proofs/writing-proofs/equality.rst

```diff
@@ -0,0 +1,1213 @@
+=========================
+Reasoning with equalities
+=========================
+
+There are multiple notions of :gdef:`equality` in Coq:
+
+- :gdef:`Leibniz equality` is the standard
+  way to define equality in Coq and the Calculus of Inductive Constructions,
+  which is in terms of a binary relation, i.e. a binary function that returns
+  a `Prop`.  The standard library
+  defines `eq` similar to this:
+
+   .. coqdoc::
+
+      Inductive eq {A : Type} (x : A) : A -> Prop := eq_refl : eq x x.
+
+  The notation `x = y` represents the term `eq x y`.  The notation `x = y :> A`
+  gives the type of x and y explicitly.
+
+- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
+  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
+  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
+  or :gdef:`quotient`
+  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
+  with setoids rather than constructing quotients, for which there is no specific support.
+
+- :gdef:`Definitional equality <definitional equality>` is equality based on the
+  :ref:`conversion rules <Conversion-rules>`, which Coq can determine automatically.
+  When two terms are definitionally equal, Coq knows it can
+  replace one with the other, such as with :tacn:`change` `X with Y`, among many
+  other advantages.  ":term:`Convertible <convertible>`" is another way of saying that
+  two terms are definitionally equal.
+
+Tactics for simple equalities
+-----------------------------
+
+.. tacn:: reflexivity
+
+   For a goal with the form :n:`{? forall @open_binders , } t = u`,
+   verifies that `t` and `u` are
+   :term:`definitionally equal <definitional equality>`, and if so,
+   solves the goal.  If not, it fails.  It is equivalent to ``apply refl_equal``.
+
+   The tactic may also be applied to goals with the form
+   :n:`{? forall @open_binders , } R @term__1 @term__2` where
+   `R` is a reflexive relation registered with the `Equivalence` or `Reflexive`
+   typeclasses.  See :cmd:`Class` and :cmd:`Instance`.
+
+   .. exn:: The relation @ident is not a declared reflexive relation. Maybe you need to require the Coq.Classes.RelationClasses library
+      :undocumented:
+
+.. tacn:: symmetry {? @simple_occurrences }
+
+   Changes a goal that has the form :n:`{? forall @open_binders , } t = u` into
+   :n:`{? forall @open_binders , } u = t`.  :n:`@simple_occurrences`
```

Comment:
Actually, in this case too, it introduces the binders before doing the symmetry, and doesn't revert the introduction afterward.